File: Java/2_String/LC_43_MultiplyStrings.java
class Solution {
    public String multiply(String num1, String num2) {
        if(num1.equals("0") || num2.equals("0")){
            return "0";
        }

        int m = num1.length();
        int n = num2.length();

        //逆序末位填充
        int[] tmp = new int[m + n];
        for(int i = m - 1; i >= 0; i--){
            int v_1 = num1.charAt(i) - '0';
            for(int j = n - 1; j >= 0; j--){
                int v_2 = num2.charAt(j) - '0';
                int sum = tmp[i + j + 1] + v_1 * v_2;
                tmp[i + j + 1] = sum % 10; //填充 i + j + 1 位
                tmp[i + j] += sum / 10;//填充 i + j 位
            }
        }

        StringBuilder res = new StringBuilder();

        //移除前置 0
        int index = 0;
        while(tmp[index] == 0){
            index++;
        }

        for(int i = index; i < tmp.length; i++){
            res.append(tmp[i]);
        }

        return res.toString();
    }
}

File: Swift/Leetcode.playground/Sources/2_String/LC_43_MultiplyStrings.swift
public class Solution_43 {
    public func multiply(_ num1: String, _ num2: String) -> String {
         if num1 == "0" || num2 == "0"{
            return "0"
        }
        
        let num_1 = Array(num1).map{ Int(String($0))! }
        let num_2 = Array(num2).map{ Int(String($0))! }

        let m = num_1.count
        let n = num_2.count

        var tmp = [Int](repeating: 0, count: m + n)

        //逆序末位填充
        for index_i in 1...m {
            let i = m - index_i
            let v_1 = num_1[i]
            for index_j in 1...n {
                let j = n - index_j
                let v_2 = num_2[j]
                let sum = tmp[i + j + 1] + v_1 * v_2
                tmp[i + j + 1] = sum % 10
                tmp[i + j] += sum / 10
            }
        }
        
        var index = 0
        while(tmp[index] == 0){
            index += 1
        }
        
        var string = ""
        for i in index..<m + n{
            string.append(String(tmp[i]))
        }
        
        return string
    }
}


public class LC_43_MultiplyStrings : Solution_43{
    public override init() {}
}
